File: cs/sorts/mergesort.cs
using System.Collections.Generic;

namespace mergesort
{
    class Program
    {   
        static List<int> Mergesort(List<int> array)
        {
            if (array.Count > 1)
            {
                var result = new List<int>();
                var first_half = new List<int>();
                var second_half = new List<int>();
                for (int i=0; i < array.Count; i++)
                {
                    if (i >= array.Count / 2)
                    {
                        second_half.Add(array[i]);
                    } else
                    {
                        first_half.Add(array[i]);
                    }
                }
                first_half = Mergesort(first_half);
                second_half = Mergesort(second_half);

                while (first_half.Count > 0 || second_half.Count > 0)
                {
                    if (first_half.Count == 0 && second_half.Count > 0)
                    {
                        result.Add(second_half[0]);
                        second_half.RemoveAt(0);
                    } else if (second_half.Count == 0 && first_half.Count > 0)
                    {
                        result.Add(first_half[0]);
                        first_half.RemoveAt(0);
                    } else if (first_half[0] < second_half[0])
                    {
                        result.Add(first_half[0]);
                        first_half.RemoveAt(0);
                    } else
                    {
                        result.Add(second_half[0]);
                        second_half.RemoveAt(0);
                    }
                }
                return result;
            }
            return array;
        }
    }
    
}

File: py/sorts/mergesort.py
# A mergesort

def mergesort(values):
    if len(values) > 1:
        first_half, second_half, result = mergesort(values[:len(values)//2]), mergesort(values[len(values)//2:]), []
        while first_half or second_half:
            if not first_half and second_half:
                result.append(second_half.pop(0))
            elif not second_half and first_half:
                result.append(first_half.pop(0))
            elif first_half[0] < second_half[0]:
                result.append(first_half.pop(0))
            else:
                result.append(second_half.pop(0))            
        return result
    return values


assert mergesort([]) == []
assert mergesort([1]) == [1], mergesort([1])
assert mergesort([1,2,3]) == [1,2,3], mergesort([1,2,3])
assert mergesort([3,2,1]) == [1,2,3]
assert mergesort([3,6,3,8,9,1,4]) == [1,3,3,4,6,8,9]
assert mergesort([3,1,5]) == [1,3,5]
assert mergesort([]) == []
assert mergesort([2,1]) == [1,2]
assert mergesort([2,1,1]) == [1,1,2]
assert mergesort([2,1,1,2,5]) == [1,1,2,2,5]